diff -up vsftpd-2.2.2/defs.h.aslim vsftpd-2.2.2/defs.h
--- vsftpd-2.2.2/defs.h.aslim	2013-03-22 14:08:32.939498577 +0100
+++ vsftpd-2.2.2/defs.h	2013-03-22 14:09:23.623498732 +0100
@@ -16,9 +16,10 @@
 #define VSFTP_LISTEN_BACKLOG    32
 #define VSFTP_SECURE_UMASK      077
 #define VSFTP_ROOT_UID          0
-/* Must be greater than both VSFTP_MAX_COMMAND_LINE and VSFTP_DIR_BUFSIZE */
-#define VSFTP_PRIVSOCK_MAXSTR   VSFTP_DATA_BUFSIZE
-#define VSFTP_AS_LIMIT          100 * 1024 * 1024
+/* Must be at least the size of VSFTP_MAX_COMMAND_LINE, VSFTP_DIR_BUFSIZE and
+   VSFTP_DATA_BUFSIZE*2 */
+#define VSFTP_PRIVSOCK_MAXSTR   VSFTP_DATA_BUFSIZE * 2
+#define VSFTP_AS_LIMIT          200UL * 1024 * 1024
 
 #endif /* VSF_DEFS_H */
 
